### PR TITLE
Configure gRPC generation

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR implements the gRPC code gen configuration.
Previously, setting the `spine.java().grpc` flag had no visible impact. When the flag is in `true`, the gRPC stubs are generated for the Protobuf services.

Also, this PR reduces the amount of additional configuration to be performed by the user by adding Spine and gRPC dependencies to the project automatically, for Java projects.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@alexander-yevsyukov, PTAL.

File: plugin/src/main/java/io/spine/tools/bootstrap/CodeGenExtension.java [hunk 1]
```diff
@@ -48,18 +58,122 @@
...
+        private ProtobufGenerator protobufGenerator;
+        private PluginTarget pluginTarget;
+        private DependencyTarget dependencyTarget;
+        private Project project;
+
+        Builder(ProtocPlugin codeGenJob) {
+            this.codeGenJob = codeGenJob;
+        }
+
+        ProtobufGenerator getProtobufGenerator() {
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/CodeGenExtension.java]: Why do we need `get` prefixes here?

Comment (alexander-yevsyukov) [reply plugin/src/main/java/io/spine/tools/bootstrap/CodeGenExtension.java]: The `set` prefix in builders look reasonable to me. But we can do without `get` in handcrafted builder classes.

File: plugin/src/main/java/io/spine/tools/bootstrap/CodeLayout.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+package io.spine.tools.bootstrap;
+
+import java.nio.file.Path;
+
+/**
+ * A description of the project code layout on the file system.
+ *
+ * <p>Typically represented by the project {@link org.gradle.api.tasks.SourceSetContainer}.
+ */
+public interface CodeLayout {
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/CodeLayout.java]: This name looks like a code formatting thing. Why isn't it `SourceSet`, for example?

File: plugin/src/main/java/io/spine/tools/bootstrap/CodeLayout.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * Marks the given directory as a generated code root dir.
+     *
+     * <p>The directory should have a sub-directory for each source set present in the project.
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/CodeLayout.java]: > The directory should have a sub-directory...

... then it's probably not `Path` but rather a custom class or interface, which reflects the fact that it has three sub-directories.

File: plugin/src/main/java/io/spine/tools/bootstrap/DependencyTarget.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+     * @see #compile(String)
+     */
+    default void compile(Artifact artifact) {
+        compile(artifact.notation());
+    }
+
+    /**
+     * Adds a new dependency within the {@code compile} configuration.
+     *
+     * <p>Though {@code compile} configuration is deprecated in Gradle, it is still used in order to
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/DependencyTarget.java]: What do we do when Gradle removes it?

File: plugin/src/main/java/io/spine/tools/bootstrap/DependencyTarget.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+     * define Protobuf dependencies without re-generating the Java/JS sources from the upstream
+     * Protobuf definitions.
+     *
+     * @see #depend(String, String)
+     */
+    default void compile(String notation) {
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/DependencyTarget.java]: The `notation` parameter needs description. Or, put the `depend(String, String)` method before those that refer to it in the `@see` clause.

File: plugin/src/main/java/io/spine/tools/bootstrap/Ext.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Accesses the project's {@code ext} values.
+ */
+final class Ext {
+
+    @SuppressWarnings("DuplicateStringLiteralInspection") // Used in other contexts.
+    private static final String SPINE_VERSION = "spineVersion";
+    private static final String GRPC = "grpc";
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/Ext.java]: Can we transform the `Property` class into enum, having `grpc`, `protoc`, `build` etc. as its members?

File: plugin/src/main/java/io/spine/tools/bootstrap/Ext.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    private Property property(String name) {
+        Object value = ext.get(name);
+        checkNotNull(value);
+        return new Property(value);
+    }
+
+    /**
+     * A project extra property.
+     */
+    private static final class Property {
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/Ext.java]: This functionality looks like a useful thing not only for bootstrapping. It is reusable enough for working with Gradle model for other things.

File: plugin/src/main/java/io/spine/tools/bootstrap/JavaExtension.java [hunk 8]
```diff
@@ -20,46 +20,136 @@
 
 package io.spine.tools.bootstrap;
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/JavaExtension.java]: I think the package should be either `io.spine.tools.gradle.bootstrap` (under `gradle`) or `io.spine.tools.gradle` (assuming the reusability) with a small fraction of code residing under `gradle.bootstrap` for the very purpose of bootstrapping.

File: plugin/src/main/java/io/spine/tools/bootstrap/protobuf/ProtocPlugin.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.tools.bootstrap.protobuf;
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/bootstrap/protobuf/ProtocPlugin.java]: How about `io.spine.tools.protoc` package for this?

**Review by alexander-yevsyukov — CHANGES_REQUESTED** [review]
Please see my comments.

**Comment by dmdashenkov** [comment]
@alexander-yevsyukov, PTAL again.

File: plugin/src/main/java/io/spine/tools/gradle/SourceLayout.java [hunk 10]
```diff
@@ -18,28 +18,20 @@
...
-import java.nio.file.Path;
+package io.spine.tools.gradle;
 
 /**
  * A description of the project code layout on the file system.
  *
  * <p>Typically represented by the project {@link org.gradle.api.tasks.SourceSetContainer}.
  */
-public interface CodeLayout {
+public interface SourceLayout {
```
Comment (alexander-yevsyukov) [inline plugin/src/main/java/io/spine/tools/gradle/SourceLayout.java]: How about `DirectoryStructure` or something like this?

**Review by alexander-yevsyukov — CHANGES_REQUESTED** [review]
Please see my comments.

**Comment by dmdashenkov** [comment]
@alexander-yevsyukov, PTAL.

**Review by alexander-yevsyukov — APPROVED** [review]
LGTM

**Pull request merged**